framework/include/base/MooseVariable.h

```diff
@@ -196,8 +196,16 @@ class MooseVariable : public MooseVariableBase
   const Node *& node() { return _node; }
   dof_id_type & nodalDofIndex() { return _nodal_dof_index; }
   bool isNodalDefined() { return _is_defined; }
-  const VariableValue & nodalSln() { return _nodal_u; }
-  const VariableValue & nodalSlnOld() { return _nodal_u_old; }
+  const VariableValue & nodalSln()
+  {
+    _need_nodal_u = true;
+    return _nodal_u;
+  }
+  const VariableValue & nodalSlnOld()
+  {
+    _need_nodal_u_old = true;
+    return _nodal_u_old;
+  }
```

Comment:
I remember `ElementKernel`, but AFAIK it is not in MOOSE right now. So if you implemented something like that in Rattlesnake, you are not fixing a bug, you are just using an incorrect API. Go look at `MooseVariable::nodalValue()` that will flip the flag and will give you the right memory back.

Or if you have a MOOSE input file that shows this is a problem, I would like to see that.